clarity_ext/repository/step_repository.py

```diff
@@ -169,3 +173,168 @@ def get_process_type(self):
+        return self
+
+    def next(self):
+        return self.__next__()
+
+    def __next__(self):
+        if self.counter == len(self.pairs):
+            raise StopIteration
+        pair = self.pairs[self.counter]
+        self.counter += 1
+        return iter((pair.input_artifact, pair.output_artifact, pair.output_generation_type))
+
+    def _assemble_pairs(self, stateful_artifacts, stateless_artifacts):
+        pairs = list()
+        stateful_by_uri = {artifact.uri: artifact for artifact in stateful_artifacts}
+        stateless_by_lims_id = {artifact.id: artifact for artifact in stateless_artifacts}
+        for input, output in self.input_output_maps:
+            stateful_input = stateful_by_uri[input["uri"].uri]
+            fetched_input = ArtifactRepresentation(
+                stateful_representation=stateful_input,
+                stateless_representation=stateless_by_lims_id[stateful_input.id]
+            )
+            stateful_output = stateful_by_uri[output["uri"].uri]
+            fetched_output = ArtifactRepresentation(
+                stateful_representation=stateful_output,
+                stateless_representation=stateless_by_lims_id[stateful_output.id]
+            )
+            pair = Pair(
+                input=fetched_input,
+                output=fetched_output,
+                output_generation_type=output["output-generation-type"],
+                state_mode=self.state_mode
+            )
+            pairs.append(pair)
+
+        return pairs
+
+    def _fetch_stateful(self):
+        artifact_keys = set()
+        for input, output in self.input_output_maps:
+            artifact_keys.add(input["uri"])
+            artifact_keys.add(output["uri"])
+
+        return self.session.api.get_batch(artifact_keys, force=self.force)
+
+    def _fetch_stateless(self):
+        artifact_keys = set()
+        for input, output in self.input_output_maps:
+            stateless_input = input["uri"].get_stateless_clone()
+            stateless_output = output["uri"].get_stateless_clone()
+            artifact_keys.add(stateless_input)
+            artifact_keys.add(stateless_output)
+
+        return self.session.api.get_batch(artifact_keys, force=self.force)
+
+
+class Pair(namedtuple('Pair', ['input', 'output', 'output_generation_type', 'state_mode'])):
+    @property
+    def input_artifact(self):
+        return self._get_artifact(self.input)
+
+    @property
+    def output_artifact(self):
+        return self._get_artifact(self.output)
+
+    def _get_artifact(self, artifact_repr):
+        if self.state_mode == WrappablePairs.MODE_STATEFUL:
+            return artifact_repr.stateful_representation
+        elif self.state_mode == WrappablePairs.MODE_STATELESS:
+            return artifact_repr.stateless_representation
+        else:
+            raise Exception("Unknown state: {}".format(self.state_mode))
+
+    def validate(self, api):
+        errors = list()
+        if not api.is_equal(
```

Comment:
My feeling is that the check for equality should only be in clarity-ext. My reasoning is that we will eventually want to send all of our patches to upstream genologics, and in that case I would prefer that we don't need to argue for any of the patches. Since we're implementing this feature in clarity-ext, it makes sense to me that we also assert that it works as expected only in clarity-ext.

This is not saying that this might not be of use in genologics also, but it's just simpler to manage the future pull requests if we leave that question out of it :)

Reply:
... but fetching an artifact with or without state seems basic enough to me for us to be able to get it into upstream eventually.

Reply:
is_equal read the xml, sort it, remove whitespaces to make it comparable with other instances. It also give the option to exclude a named tag (qc-flag). I didn't want to include it in clarity-ext because it's so very much about the internal of a genologics entity, (the xml). Apart from we are two groups, I think there's no doubt this function belongs to the genologics repo.

Reply:
Yes, I think I thought of this when I wrote the code, where to put it. I still think it belongs to the genologics repo, but if you have a strong opinion about this, I'll move it. I'm against the existence of a "genologics bubble" within clarity-ext. In the long run that could cause problems. I also think is_equal is rather innocent, because they simply don't reference it. Also, in case they want to merge with ours repo, wouldn't that be because they want our changes. I don't see that we have to argue about that?

Reply:
There are always politics in pull requests. I think they will want most of our patches, but in my experience (with any PR to public repos) is that they will also not want a more complex codebase unless they understand the need for it for the whole.

I'm just assuming that an equality check for the XML responses is not something they will see the need for. I imagine them asking why we can't just do that at another level. But I might be wrong!

An alternative that would also work is just to remove this equality check in genologics when we remove our use of it in clarity-ext. :thinking:

I'm fine with both approaches.

Reply:
If you use that xml diff engine (https://pypi.org/project/xmldiff/) the patch will be minimal and I think they might accept it (although then they might ask us why we're adding a new pip requirement for this :)).

Reply:
And I also want to just say that I also think that `is_equal` is as you say very innocent - and there are good arguments for including it. It's definitely not worth worrying too much about at least - we can always fix it if they don't want it.

Reply:
Ok, I'll move it to the clarity-ext. I also check out the xmldiff.